Server v0.3.0-rc2: panic: runtime error: invalid memory address or nil pointer dereference

```
    2016/02/24 16:06:16 [DEBUG] nomad.heartbeat: node '3cab6545-96ed-1bd0-b369-629ed0b30da8' TTL expired
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x50 pc=0x6e7566]

goroutine 3373 [running]:
panic(0x1081960, 0xc820014080)
        /usr/lib/go/src/runtime/panic.go:464 +0x3e6
github.com/hashicorp/nomad/nomad.(*Node).createNodeEvals(0xc820230140, 0xc8203a1a40, 0x24, 0x4f7, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /go/src/github.com/hashicorp/nomad/nomad/node_endpoint.go:632 +0x1046
github.com/hashicorp/nomad/nomad.(*Node).UpdateStatus(0xc820230140, 0xc8203a1230, 0xc820136320, 0x0, 0x0)
        /go/src/github.com/hashicorp/nomad/nomad/node_endpoint.go:189 +0x93d
github.com/hashicorp/nomad/nomad.(*Server).invalidateHeartbeat(0xc820060a80, 0xc8203a1a40, 0x24)
        /go/src/github.com/hashicorp/nomad/nomad/heartbeat.go:101 +0x3d7
github.com/hashicorp/nomad/nomad.(*Server).resetHeartbeatTimerLocked.func1()
        /go/src/github.com/hashicorp/nomad/nomad/heartbeat.go:77 +0x3f
created by time.goFunc
        /usr/lib/go/src/time/sleep.go:129 +0x3a
```

--- a/nomad/structs/structs.go
+++ b/nomad/structs/structs.go
@@ -2421,7 +2421,15 @@ type Plan struct {
 func (p *Plan) AppendUpdate(alloc *Allocation, status, desc string) {
 	newAlloc := new(Allocation)
 	*newAlloc = *alloc
-	newAlloc.Job = nil // Normalize the job
+
+	// If the job is not set in the plan we are deregistering a job so we
+	// extract the job from the allocation.
+	if p.Job == nil && newAlloc.Job != nil {
+		p.Job = newAlloc.Job
+	}
+
+	// Normalize the job
+	newAlloc.Job = nil
 	newAlloc.DesiredStatus = status
 	newAlloc.DesiredDescription = desc
 	node := alloc.NodeID

--- a/scheduler/generic_sched_test.go
+++ b/scheduler/generic_sched_test.go
@@ -634,6 +634,13 @@ func TestServiceSched_JobDeregister(t *testing.T) {
 	out, err := h.State.AllocsByJob(job.ID)
 	noErr(t, err)
 
+	// Ensure that the job field on the allocation is still populated
+	for _, alloc := range out {
+		if alloc.Job == nil {
+			t.Fatalf("bad: %#v", alloc)
+		}
+	}
+
 	// Ensure no remaining allocations
 	out = structs.FilterTerminalAllocs(out)
 	if len(out) != 0 {

